src/base/utils/version.h

```diff
@@ -0,0 +1,185 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#ifndef QBITTORRENT_UTILS_VERSION_H
+#define QBITTORRENT_UTILS_VERSION_H
+
+#include <array>
+#include <stdexcept>
+
+#include <QDebug>
+#include <QString>
+#include <QStringList>
+
+namespace Utils
+{
+
+    template <class T, std::size_t N, std::size_t Mandatory = N>
+    class Version
+    {
+        static_assert(N > 0, "The number of version components may not be smaller than 1");
+        static_assert(N >= Mandatory,
+                      "The number of mandatory components may not be larger than the total number of components");
+
+    public:
+        typedef T ComponentType;
+        typedef Version<T, N, Mandatory> ThisType;
+
+        constexpr Version()
+            : m_components {}
+        {
+        }
+
+        constexpr Version(const ThisType &other) = default;
+
+        template <class ... Other>
+        constexpr Version(Other ... components)
+            : m_components {{components ...}}
+        {
+        }
+
+        Version(const QString &version)
+            : Version {version.split(QLatin1Char('.'))}
+        {
+        }
+
+        Version(const QByteArray &version)
+            : Version {version.split('.')}
+        {
+        }
+
+        ComponentType majorNumber() const
+        {
+            static_assert(N >= 1, "The number of version components is too small");
+            return (*this)[0];
+        }
+
+        ComponentType minorNumber() const
+        {
+            static_assert(N >= 2, "The number of version components is too small");
+            return (*this)[1];
+        }
+
+        ComponentType revisionNumber() const
+        {
+            static_assert(N >= 3, "The number of version components is too small");
+            return (*this)[2];
+        }
+
+        ComponentType patchNumber() const
+        {
+            static_assert(N >= 4, "The number of version components is too small");
+            return (*this)[3];
+        }
+
+        ComponentType operator[](std::size_t i) const
+        {
+            return m_components.at(i);
+        }
+
+        operator QString() const
```

Comment:
I am not sure the behavior of this is desired. If someone enters a version of "3.0.0" then he surely expects "3.0.0" printed back and not just "3". Even if "Mandatory == 1".

Do you have any use-case that need the current behavior?